### DISCUSSION TITLE
SYNTH_HIERARCHICAL=1 running times and what does it mean?

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD-flow-scripts#1647] [General] [answered]
I've set  `SYNTH_HIERARCHICAL=1` because in `RTLMP=True` examples it is set.

What does `SYNTH_HIERARCHICAL=1` do?

I know it does "hierarchical synthesis instead of flattened", but a bit more elaboration on what this means and what to expect  would be helpful.

Is it expected that `1_1_yosys_hier_report` takes as long as it does below on megaboom top level?

```
Log                       Elapsed seconds
1_1_yosys                       2768
1_1_yosys_hier_report           1486
```

**Comment by maliberty** [comment, marked as answer]
Synthesis goes module by module but does not optimize across module boundaries.  If you flatten then you have a single module so you get maximum optimization but the largest runtime.   You also lose that information that is useful for mpl2 and the user.

If you go with purely hierarchical the runtime is better but the results worse.  Ideally you want to flatten small modules and keep larger ones.    Its hard to know do that with yosys today.  So we run a purely hierarchical run, get the sizes, build a list of modules to keep, and re-run yosys.  @gadfort has some suggestions on how we might mitigate the cost but I haven't had time to invest in exploring them.
